File: libs/src/java/salmon-fs/com/mku/fs/file/HttpFile.java
package com.mku.fs.file;
/*
MIT License

Copyright (c) 2025 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import com.mku.fs.streams.HttpFileStream;
import com.mku.salmon.encode.Base64Utils;
import com.mku.streams.MemoryStream;
import com.mku.streams.RandomAccessStream;

import java.io.Closeable;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * IFile implementation for a remote HTTP file.
 */
public class HttpFile implements IFile {
    /**
     * Directory separator
     */
    public static final String separator = "/";
    private static final String DATE_FORMAT = "EEE, dd MMM yyyy HH:mm:ss zzz";

    private String filePath;
    private HttpFile.Response response;


	/**
	 * Get the credentials
	 * @return The credentials
	 */
    public Credentials getCredentials() {
        return credentials;
    }

    private Credentials credentials;

	/**
	 * Set the user credentials
	 * @param credentials The credentials
	 */
    public void setCredentials(Credentials credentials) {
        this.credentials = credentials;
    }

    /**
     * Instantiate a real file represented by the filepath provided (Remote read-write drive)
     *
     * @param path The filepath. This should be a relative path of the vault folder
     */
    public HttpFile(String path) {
        this(path, null);
    }
    /**
     * Instantiate a real file represented by the filepath provided (Remote read-write drive)
     *
     * @param path The filepath. This should be a relative path of the vault folder
	 * @param credentials The credentials (Basic Auth)
     */
    public HttpFile(String path, Credentials credentials) {
        this.filePath = path;
		this.credentials = credentials;
    }

    private synchronized HttpFile.Response getResponse() throws Exception {
        if(this.response == null) {
			HttpURLConnection conn = null;
            try {
                // TODO: should this be INFO method?
                conn = createConnection("GET", this.filePath);
                setDefaultHeaders(conn);
				setServiceAuth(conn);
                conn.connect();
                checkStatus(conn, HttpURLConnection.HTTP_OK);
				this.response = new Response(conn.getHeaderFields());
            } finally {
                closeConnection(conn);
            }
        }
        return this.response;
    }

    /**
     * Create a directory under this directory.
     *
     * @param dirName The name of the new directory.
     * @return The newly created directory.
     */
    public IFile createDirectory(String dirName) {
        throw new UnsupportedOperationException("Unsupported Operation, readonly filesystem");
    }

    /**
     * Create a file under this directory.
     *
     * @param filename The name of the new file.
     * @return The newly created file.
     * @throws IOException Thrown if there is an IO error.
     */
    public IFile createFile(String filename) throws IOException {
        throw new UnsupportedOperationException("Unsupported Operation, readonly filesystem");
    }

    /**
     * Delete this file or directory.
     *
     * @return True if deletion is successful.
     */
    public boolean delete() {
        throw new UnsupportedOperationException("Unsupported Operation, readonly filesystem");
    }

    /**
     * Check if file or directory exists.
     *
     * @return True if exists.
     */
    public boolean exists() {
        try {
            return this.getResponse() != null;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Get the display path on the physical disk.
     *
     * @return The display path.
     */
    public String getDisplayPath() {
        return filePath;
    }

    /**
     * Get the name of this file or directory.
     *
     * @return The name of this file or directory.
     */
    public String getName() {
        if (this.filePath == null)
            throw new RuntimeException("Filepath is not assigned");
        String nFilePath = this.filePath;
        if (nFilePath.endsWith(HttpFile.separator))
            nFilePath = nFilePath.substring(0, nFilePath.length() - 1);
        String[] parts = nFilePath.split(HttpFile.separator);
        String basename = parts[parts.length - 1];
        if (basename == null)
            throw new RuntimeException("Could not get the file name");
        if (basename.contains("%")) {
			try {
				// do not use the charset variable for backwards compatibility with android
				basename = URLDecoder.decode(basename, StandardCharsets.UTF_8.name());
			} catch (Exception e) {
                throw new RuntimeException(e);
            } 
        }
        return basename;
    }

    /**
     * Get a stream for reading the file.
     *
     * @return The stream to read from.
     * @throws FileNotFoundException Thrown if file not found
     */
    public RandomAccessStream getInputStream() throws FileNotFoundException {
        return new HttpFileStream(this, "r");
    }

    /**
     * Get a stream for writing to this file.
     *
     * @return The stream to write to.
     * @throws FileNotFoundException Thrown if file not found
     */
    public RandomAccessStream getOutputStream() throws FileNotFoundException {
        throw new UnsupportedOperationException("Unsupported Operation, readonly filesystem");
    }

    /**
     * Get the parent directory of this file or directory.
     *
     * @return The parent directory.
     */
    public IFile getParent() {
        if (filePath.length() == 0 || filePath.equals(HttpFile.separator))
            return null;
        String path = filePath;
        if (path.endsWith(HttpFile.separator))
            path = path.substring(0, path.length() - 1);
        int index = path.lastIndexOf(HttpFile.separator);
        if (index == -1)
            return null;
        HttpFile parent = new HttpFile(path.substring(0, index), credentials);
        return parent;
    }

    /**
     * Get the path of this file.
     *
     * @return The path
     */
    public String getPath() {
        return filePath;
    }

    /**
     * Check if this is a directory.
     *
     * @return True if it's a directory.
     */
    public boolean isDirectory() {
        HttpFile.Response res = null;
        try {
            res = this.getResponse();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        if (res == null)
            throw new RuntimeException("Could not get response");
        String contentType = res.getHeader("Content-Type");
        if (contentType == null)
            throw new RuntimeException("Could not get content type");
        return contentType.startsWith("text/html");
    }

    /**
     * Check if this is a file.
     *
     * @return True if it's a file
     */
    public boolean isFile() {
        return !isDirectory() && exists();
    }

    /**
     * Get the last modified date on disk.
     *
     * @return The last modified date in milliseconds
     */
    public long getLastDateModified() {
        String lastDateModified = null;
        try {
            lastDateModified = getResponse().getHeader("last-modified");
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        if (lastDateModified == null) {
            lastDateModified = "0";
        }
        Date date = null;
        try {
            date = new SimpleDateFormat(DATE_FORMAT).parse(lastDateModified);
        } catch (ParseException e) {
            throw new RuntimeException(e);
        }
        long lastModified = date.getTime() / 1000;
        return lastModified;
    }

    /**
     * Get the size of the file on disk.
     *
     * @return The length
     */
    public long getLength() {
        HttpFile.Response res;
        try {
            res = this.getResponse();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        if (res == null)
            throw new RuntimeException("Could not get response");

        long length = 0;
        String lenStr = res.getHeader("content-length");
        if (lenStr != null) {
            length = Integer.parseInt(lenStr);
        }
        return length;
    }

    /**
     * Get the count of files and subdirectories
     *
     * @return The children count
     */
    public int getChildrenCount() {
        return this.listFiles().length;
    }

    /**
     * List all files under this directory.
     *
     * @return The list of files.
     */
    public IFile[] listFiles() {
        List<IFile> files = new ArrayList<>();
        if (this.isDirectory()) {
            RandomAccessStream stream = null;
            MemoryStream ms = null;
            try {
                stream = this.getInputStream();
                ms = new MemoryStream();
                stream.copyTo(ms);
                String contents = new String(ms.toArray());

                Matcher matcher = Pattern.compile("(.+?)HREF=\"(.+?)\"(.+?)", Pattern.CASE_INSENSITIVE).matcher(contents);
                while (matcher.find()) {
                    String filename = matcher.group(2);
                    if (filename.contains(":") || filename.contains(".."))
                        continue;
                    if (filename.contains("%")) {
						// do not use the charset variable for backwards compatibility with android
                        filename = URLDecoder.decode(filename, StandardCharsets.UTF_8.name());
                    }
                    IFile file = new HttpFile(this.filePath + HttpFile.separator + filename, credentials);
                    files.add(file);
                }
            } catch (Exception e) {
                throw new RuntimeException(e);
            } finally {
                if (ms != null)
                    ms.close();
                if (stream != null) {
                    try {
                        stream.close();
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                }
            }
        }
        return files.toArray(new IFile[0]);
    }

    /**
     * Move this file or directory under a new directory. Not supported.
     *
     * @param newDir  The target directory.
     * @return The moved file. Use this file for subsequent operations instead of the original.
     */
    public IFile move(IFile newDir) throws IOException {
        return move(newDir, null);
    }

    /**
     * Move this file or directory under a new directory. Not supported.
     *
     * @param newDir  The target directory.
     * @param options The options
     * @return The moved file. Use this file for subsequent operations instead of the original.
     */
    public IFile move(IFile newDir, MoveOptions options) throws IOException {
        throw new UnsupportedOperationException("Unsupported Operation, readonly filesystem");
    }

    /**
     * Move this file or directory under a new directory. Not supported.
     *
     * @param newDir The target directory.
     * @return The copied file. Use this file for subsequent operations instead of the original.
     * @throws IOException Thrown if there is an IO error.
     */
    @Override
    public IFile copy(IFile newDir) throws IOException {
        return copy(newDir, null);
    }

    /**
     * Move this file or directory under a new directory. Not supported.
     *
     * @param newDir           The target directory.
     * @param options          The options
     * @return The copied file. Use this file for subsequent operations instead of the original.
     * @throws IOException Thrown if there is an IO error.
     */
    @Override
    public IFile copy(IFile newDir, CopyOptions options) throws IOException {
        throw new UnsupportedOperationException("Unsupported Operation, readonly filesystem");
    }

    /**
     * Get the file or directory under this directory with the provided name.
     *
     * @param filename The name of the file or directory.
     * @return The child
     */
    public IFile getChild(String filename) {
        if (isFile())
            return null;
        String nFilepath = this.getChildPath(filename);
        HttpFile child = new HttpFile(nFilepath, credentials);
        return child;
    }

    /**
     * Rename the current file or directory.
     *
     * @param newFilename The new name for the file or directory.
     * @return True if successfully renamed.
     */
    public boolean renameTo(String newFilename) {
        throw new UnsupportedOperationException("Unsupported Operation, readonly filesystem");
    }

    /**
     * Create this directory under the current filepath.
     *
     * @return True if created.
     */
    public boolean mkdir() {
        throw new UnsupportedOperationException("Unsupported Operation, readonly filesystem");
    }

    /**
     * Reset cached properties
     */
    public void reset() {
        this.response = null;
    }

    private String getChildPath(String filename) {
        String nFilepath = this.filePath;
        if (!nFilepath.endsWith(HttpFile.separator))
            nFilepath += HttpFile.separator;
        nFilepath += filename;
        return nFilepath;
    }

    /**
     * Returns a string representation of this object
     */
    @Override
    public String toString() {
        return filePath;
    }


    private HttpURLConnection createConnection(String method, String url) throws IOException {
        HttpURLConnection conn = HttpSyncClient.createConnection(url);
        conn.setUseCaches(false);
        conn.setDefaultUseCaches(false);
        conn.setRequestMethod(method);
        conn.setDoInput(true);
        return conn;
    }

    private void closeConnection(HttpURLConnection conn) {
        if (conn != null) {
            try {
                conn.disconnect();
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
    }

    private void closeStream(Closeable stream) {
        if(stream != null) {
            try {
                stream.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }

	private void setServiceAuth(HttpURLConnection conn) {
		if(credentials != null) {
			String encoding = Base64Utils.getBase64().encode((credentials.getServiceUser()
					+ ":" + credentials.getServicePassword()).getBytes());
			conn.setRequestProperty("Authorization", "Basic " + encoding);
		}
    }
	
    private void checkStatus(HttpURLConnection conn, int status) throws IOException {
        if (conn.getResponseCode() != status)
            throw new IOException(conn.getResponseCode()
                    + " " + conn.getResponseMessage());
    }

    private void setDefaultHeaders(HttpURLConnection conn) {
        conn.setRequestProperty("Cache", "no-store");
    }
	
    private static class Response {
        private Map<String, List<String>> headers;

        Response(Map<String, List<String>> headers) {
            this.headers = headers;
        }
		
		String getHeader(String name) {
			for(Map.Entry<String, List<String>> entry : headers.entrySet()){
				// status has a null key
				if(entry.getKey() == null)
					continue;
				if(entry.getKey().toLowerCase().equals(name.toLowerCase())) {
					if(entry.getValue().size() > 0) {
						return entry.getValue().get(0);
					} else {
						return null;
					}
				}
			}
			return null;
		}
    }
}

File: libs/src/typescript/salmon-fs/fs/file/http_file.ts
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import { RandomAccessStream } from '../../../salmon-core/streams/random_access_stream.js';
import { CopyOptions, IFile, MoveOptions } from './ifile.js';
import { HttpFileStream } from '../streams/http_file_stream.js';
import { IOException } from '../../../salmon-core/streams/io_exception.js';
import { MemoryStream } from '../../../salmon-core/streams/memory_stream.js';

/**
 * Salmon RealFile implementation for Javascript.
 */
export class HttpFile implements IFile {
	/**
	 * Directory separator
	 */
    public static readonly separator: string = "/";

    filePath: string;
    response: Response | null = null;

    /**
     * Instantiate a real file represented by the filepath provided.
     * @param {string} path The filepath.
     */
    public constructor(path: string) {
        this.filePath = path;
    }

    async #getResponse(): Promise<Response> {
        if (this.response == null) {
			let headers = new Headers();
			this.#setDefaultHeaders(headers);
            this.response = (await fetch(this.filePath, 
			{method: 'HEAD', keepalive: true, headers: headers}));
			await this.#checkStatus(this.response, 200);
		}
        return this.response;
    }

    /**
     * Create a directory under this directory.
     * @param {string} dirName The name of the new directory.
     * @returns The newly created directory.
     */
    public async createDirectory(dirName: string): Promise<IFile> {
        throw new Error("Unsupported Operation, readonly filesystem");
    }

    /**
     * Create a file under this directory.
     * @param {string} filename The name of the new file.
     * @returns {Promise<IFile>} The newly created file.
     * @throws IOException Thrown if there is an IO error.
     */
    public createFile(filename: string): Promise<IFile> {
        throw new Error("Unsupported Operation, readonly filesystem");
    }

    /**
     * Delete this file or directory.
     * @returns {Promise<boolean>} True if deletion is successful.
     */
    public async delete(): Promise<boolean> {
        throw new Error("Unsupported Operation, readonly filesystem");
    }

    /**
     * True if file or directory exists.
     * @returns {Promise<boolean>} True if exists
     */
    public async exists(): Promise<boolean> {
        return (await this.#getResponse()).status == 200 || (await this.#getResponse()).status == 206;
    }

    /**
     * Get the path of this file. For Javascript this is the same as the absolute filepath.
     * @returns {string} The path
     */
    public getPath(): string {
        return this.filePath;
    }

    /**
     * Get the absolute path on the physical disk. For javascript this is the same as the filepath.
     * @returns {string} The absolute path.
     */
    public getDisplayPath(): string {
        return this.filePath;
    }

    /**
     * Get the name of this file or directory.
     * @returns {string} The name of this file or directory.
     */
    public getName(): string {
        if (this.filePath == null)
            throw new Error("Filepath is not assigned");
        let nFilePath = this.filePath;
        if(nFilePath.endsWith("/"))
            nFilePath = nFilePath.substring(0,nFilePath.length-1);
        let basename: string | undefined = nFilePath.split(HttpFile.separator).pop();
        if (basename === undefined)
            throw new Error("Could not get basename");
        if (basename.includes("%")){
            basename = decodeURIComponent(basename);
        }
        return basename;
    }

    /**
     * Get a stream for reading the file.
     * @returns {Promise<RandomAccessStream>} The stream to read from.
     * @throws FileNotFoundException
     */
    public async getInputStream(): Promise<RandomAccessStream> {
        let fileStream: HttpFileStream = new HttpFileStream(this, "r");
        return fileStream;
    }

    /**
     * Get a stream for writing to this file.
     * @returns {Promise<RandomAccessStream>} The stream to write to.
     * @throws FileNotFoundException
     */
    public async getOutputStream(): Promise<RandomAccessStream> {
        throw new Error("Unsupported Operation, readonly filesystem");
    }

    /**
     * Get the parent directory of this file or directory.
     * @returns {Promise<IFile>} The parent directory.
     */
    public async getParent(): Promise<IFile> {
		let path: string = this.filePath;
		if(path.endsWith(HttpFile.separator))
			path = path.slice(0,-1);
        let parentFilePath: string = path.substring(0, path.lastIndexOf(HttpFile.separator));
        return new HttpFile(parentFilePath);
    }

    /**
     * True if this is a directory.
     * @returns {Promise<boolean>} True if directory
     */
    public async isDirectory(): Promise<boolean> {
        let res: Response = (await this.#getResponse());
        if (res == null)
            throw new Error("Could not get response");
        if (res.headers == null)
            throw new Error("Could not get headers");
        let contentType: string | null = res.headers.get("Content-Type");
        if (contentType == null)
            throw new Error("Could not get content type");
        return contentType.startsWith("text/html");
    }

    /**
     * True if this is a file.
     * @returns {Promise<boolean>} True if file
     */
    public async isFile(): Promise<boolean> {
        return !await this.isDirectory() && await this.exists();
    }

    /**
     * Get the last modified date on disk.
     * @returns {Promise<number>} The last date modified
     */
    public async getLastDateModified(): Promise<number> {
        let headers: Headers = (await this.#getResponse()).headers;
        let lastDateModified: string | null = headers.get("last-modified");
        if (lastDateModified == null) {
			lastDateModified = headers.get("date");
		}
		if (lastDateModified == null) {
			lastDateModified = "0";
		}
        let date: Date = new Date(lastDateModified);
        let lastModified = date.getTime();
        return lastModified;
    }

    /**
     * Get the size of the file on disk.
     * @returns {Promise<number>} The size
     */
    public async getLength(): Promise<number> {
        let res: Response = (await this.#getResponse());
        if (res == null)
            throw new IOException("Could not get response");

        let length: number = 0;
        let lenStr: string | null = res.headers.get("content-length");
        if (lenStr) {
            length = parseInt(lenStr);
        }
        return length;
    }

    /**
     * Get the count of files and subdirectories
     * @returns {Promise<number>} The number of files and subdirectories
     */
    public async getChildrenCount(): Promise<number> {
        return (await this.listFiles()).length;
    }
    /**
     * List all files under this directory.
     * @returns {Promise<IFile[]>} The list of files.
     */
    public async listFiles(): Promise<IFile[]> {
		if(await this.isDirectory()) {
			let files: IFile[] = [];
			let stream: RandomAccessStream = await this.getInputStream();
			let ms: MemoryStream = new MemoryStream();
			await stream.copyTo(ms);
			await ms.close();
			await stream.close();
			let contents: string = new TextDecoder().decode(ms.toArray());
			let matches = contents.matchAll(/HREF\=\"(.+?)\"/ig);
			for (const match of matches) {
				let filename: string = match[1];
				if (filename.includes(":") || filename.includes(".."))
					continue;
				if (filename.includes("%")){
					filename = decodeURIComponent(filename);
				}
				let file: IFile = new HttpFile(this.filePath + HttpFile.separator + filename);
				files.push(file);
			}
			return files;
		}
		return [];
    }

    /**
     * Move this file or directory under a new directory. Not supported.
     * @param {IFile} newDir The target directory.
     * @param {MoveOptions} [options] The options
     * @returns {Promise<IFile>} The moved file. Use this file for subsequent operations instead of the original.
     */
    public async move(newDir: IFile, options?: MoveOptions): Promise<IFile> {
        throw new Error("Unsupported Operation, readonly filesystem");
    }

    /**
     * Move this file or directory under a new directory. Not supported.
     * @param {IFile} newDir    The target directory.
     * @param {CopyOptions} [options] The options.
     * @returns {Promise<IFile>} The copied file. Use this file for subsequent operations instead of the original.
     * @throws IOException Thrown if there is an IO error.
     */
    public async copy(newDir: IFile, options?: CopyOptions): Promise<IFile> {
        throw new Error("Unsupported Operation, readonly filesystem");
    }

    /**
     * Get the file or directory under this directory with the provided name.
     * @param {string} filename The name of the file or directory.
     * @returns {Promise<IFile | null>} The child
     */
    public async getChild(filename: string): Promise<IFile | null> {
        if (await this.isFile())
            return null;
        let child: HttpFile = new HttpFile(this.filePath + HttpFile.separator + filename);
        return child;
    }

    /**
     * Rename the current file or directory. Not supported.
     * @param {string} newFilename The new name for the file or directory.
     * @returns {Promise<boolean>} True if successfully renamed.
     */
    public async renameTo(newFilename: string): Promise<boolean> {
        throw new Error("Unsupported Operation, readonly filesystem");
    }

    /**
     * Create this directory under the current filepath. Not supported.
     * @returns {Promise<boolean>} True if created.
     */
    public async mkdir(): Promise<boolean> {
        throw new Error("Unsupported Operation, readonly filesystem");
    }
	
	/**
     * Reset cached properties
     */
    public reset() {
		this.response = null;
	}

    /**
     * Returns a string representation of this object
     * @returns {string} The string
     */
    public toString(): string {
        return this.filePath;
    }
	
	async #checkStatus(httpResponse: Response, status: number) {
        if (httpResponse.status != status)
            throw new IOException(httpResponse.status
                    + " " + httpResponse.statusText);
    }

    #setDefaultHeaders(headers: Headers) {
        headers.append("Cache", "no-store");
		headers.append("Connection", "keep-alive");
    }
}
